resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ResourceTypeValidator.scala

```diff
@@ -0,0 +1,162 @@
+    }
+
+    var customResourceTypes: Map[String, String] = null
+    (role, mode, resourceType) match {
+      case ("executor", _, _) => customResourceTypes = requestedResources
+        .customResourceTypesForExecutor
+      case ("driver", "client", _) => customResourceTypes = requestedResources
+        .customResourceTypesForDriverClientMode
+      case ("driver", "cluster", _) => customResourceTypes = requestedResources
+        .customResourceTypesForDriverClusterMode
+    }
+
+    var resourceTypeObj: String = null
+    (role, resourceType) match {
+      case ("driver", "cores") => resourceTypeObj = requestedResources.driverCores
+      case ("driver", "memory") => resourceTypeObj = requestedResources.driverMemory
+      case ("executor", "cores") => resourceTypeObj = requestedResources.executorCores
+      case ("executor", "memory") => resourceTypeObj = requestedResources.executorMemory
+    }
+
+    val (standardResourceTypeId: String, customResourceTypeId: String) =
+      getResourceTypeIdsByRole(role, mode, resourceType)
+
+    if (resourceTypeObj != null && customResourceTypes.contains(customResourceTypeId)) {
+      return formatDuplicateResourceTypeErrorMessage(standardResourceTypeId, customResourceTypeId)
+    }
+    ""
+  }
+
+  private def formatDuplicateResourceTypeErrorMessage(standardResourceTypeId: String,
+                                                      customResourceTypeId: String): String = {
+    s"$standardResourceTypeId and $customResourceTypeId" +
+      " configs are both present, only one of them is allowed at the same time!"
+  }
+
+  private def getResourceTypeIdsByRole(role: String, mode: String, resourceType: String) = {
+    val standardResourceTypeId: String = s"spark.$role.$resourceType"
+
+    var customResourceTypeId: String = ""
+    (role, mode) match {
+      case ("driver", "client") => customResourceTypeId += "spark.yarn.am.resource."
+      case ("driver", "cluster") => customResourceTypeId += "spark.yarn.driver.resource."
+      case ("executor", _) => customResourceTypeId += "spark.yarn.executor.resource."
+    }
+
+    customResourceTypeId += resourceType
+
+    (standardResourceTypeId, customResourceTypeId)
+  }
+
+  private class ResourceTypeConfigProperties(val role: String,
+                                             val mode: String = "",
+                                             val resourceType: String)
+
+
+  private class RequestedResources(val sparkConf: SparkConf) {
+    val driverMemory: String = safelyGetFromSparkConf(sparkConf, "spark.driver.memory")
+    val driverCores: String = safelyGetFromSparkConf(sparkConf, "spark.driver.cores")
+    val executorMemory: String = safelyGetFromSparkConf(sparkConf, "spark.executor.memory")
+    val executorCores: String = safelyGetFromSparkConf(sparkConf, "spark.executor.cores")
+    val customResourceTypesForDriverClientMode: Map[String, String] =
```

Comment:
I think you're a bit confused about what this is controlling. You can't request driver resources in client mode because the client does not run inside YARN. I think what you mean instead of "driver" in this case is what the rest of the code refers to as "AM".

Reply:
Thanks for pointing this. Indeed, after I read your comment and the code / Spark docs again, my understanding was not entirely correct.
I modified the code accordingly, please check again.